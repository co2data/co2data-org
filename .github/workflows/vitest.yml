name: Unit tests
on: push

concurrency: ${{ github.workflow }}-${{ github.ref }}

jobs:
  unit-test:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
      - uses: pnpm/action-setup@v2
        with:
          version: 8
      - uses: actions/setup-node@v3
        with:
          node-version: 20
          cache: 'pnpm'

      - run: pnpm install

      - name: Run unit tests
        run: pnpm test
