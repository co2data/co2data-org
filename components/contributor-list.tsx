'use client'

import { useState } from 'react'
import Co2Average from './co2-average'

export default function ContributorList({
  co2Averages,
}: {
  co2Averages: Co2Average[]
}) {
  const [term, setTerm] = useState('')
  return (
    <div className={term ? 'space-y-4' : 'space-y-20 sm:space-y-32 '}>
      <div className="space-y-12 text-center sm:space-y-20 ">
        <h1 className="text-6xl font-bold text-sky-600">
          CO<sub>2</sub> Data
        </h1>
        <div className="">
          <label className="sr-only" htmlFor="search">
            Search
          </label>
          <input
            className="rounded border-2 border-sky-600 focus:ring-4 focus:ring-sky-100"
            type="search"
            name="search"
            id="search"
            onChange={(e) => setTerm(e.target.value)}
            placeholder="Search"
            value={term}
          />
        </div>
      </div>
      <ul
        className={
          term
            ? 'flex flex-wrap justify-center gap-1'
            : 'grid grid-cols-[repeat(auto-fill,_minmax(14rem,_1fr))] gap-8'
        }
      >
        {co2Averages
          .filter((co2) =>
            term ? co2.title.toLowerCase().includes(term.toLowerCase()) : true
          )
          .map((co2Average) => (
            <Co2Average key={co2Average.slug} co2Average={co2Average} />
          ))}
      </ul>
    </div>
  )
}
