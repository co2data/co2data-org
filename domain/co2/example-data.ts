import { Co2Average } from '.'

export function makeCo2Average() {
  return {
    id: 'id',
    title: 'Pork',
    slug: 'test',
    unit: 'kilogram',
    avgPerYear: 500000,
    avgPerUnit: 30000,
    singleConsumptionFrom: 0,
    singleConsumptionAverage: 4,
    singleConsumptionTo: 100,
    timesPerYearFrom: 0,
    timesPerYearAverage: 5,
    timesPerYearTo: 20,
  } satisfies Co2Average
}
