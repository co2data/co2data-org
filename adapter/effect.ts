import { Effect, Layer, flow } from 'effect'
import { Co2Repository } from './co2-repository'
import { SourceRepository, SourceRepositoryLive } from './source-repository'
import { NodeSdkLive } from '@/infra/tracing/NodeSdkLive'
import { sourceRepoLive } from '@/domain/source'
import { co2RepoLive } from '@/domain/co2'

export function run<P, Q, A>(
  effect: (
    arg1: P,
    arg2: Q
  ) => Effect.Effect<Co2Repository | SourceRepository, never, A>
) {
  return flow(
    effect,
    Effect.provide(Layer.merge(co2RepoLive, sourceRepoLive)),
    Effect.provide(NodeSdkLive),
    Effect.runPromise
  )
}
