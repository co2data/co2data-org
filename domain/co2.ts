import { Co2RepositoryLive } from '@/adapter/co2-repository'
import { DbLive } from '@/infra/db'
import { Layer } from 'effect'
export { Co2Repository } from '@/adapter/co2-repository'
export type Unit =
  | 'gram'
  | 'hour'
  | 'kilogram'
  | 'kilometer'
  | 'liter'
  | 'meter'
  | 'minute'

export type Co2Average = {
  id: string
  title: string
  slug: string
  unit: Unit
  avgPerYear: number
  avgPerUnit: number
  singleConsumptionFrom: number
  singleConsumptionTo: number
  singleConsumptionAverage: number
  timesPerYearFrom: number
  timesPerYearTo: number
  timesPerYearAverage: number
}

export const repositoryLive = Co2RepositoryLive.pipe(Layer.use(DbLive))
