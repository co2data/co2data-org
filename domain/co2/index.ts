import { Option } from 'effect'

export { Co2Repository } from '@/domain/co2/repository'
export type Unit =
  | 'gram'
  | 'hour'
  | 'kilogram'
  | 'kilometer'
  | 'liter'
  | 'meter'
  | 'minute'

export type Co2Average = {
  id: string
  title: string
  description: Option.Option<string>
  slug: string
  unit: Unit
  avgPerYear: number
  avgPerUnit: number
  singleConsumptionFrom: number
  singleConsumptionTo: number
  singleConsumptionAverage: number
  timesPerYearFrom: number
  timesPerYearTo: number
  timesPerYearAverage: number
}
