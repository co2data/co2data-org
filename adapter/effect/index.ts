import { NodeSdkLive } from '@/adapter/tracing/NodeSdkLive'
import { Effect, flow } from 'effect'
import 'server-only'
import { Co2Repository } from '../../domain/co2/repository'
import { SourceRepository } from '../../domain/source/repository'
import { mainLive } from './main'

export function run<P, Q, A>(
  effect: (
    arg1: P,
    arg2: Q
  ) => Effect.Effect<Co2Repository | SourceRepository, never, A>
) {
  return flow(
    effect,
    Effect.provide(mainLive),
    Effect.provide(NodeSdkLive),
    Effect.runPromise
  )
}
