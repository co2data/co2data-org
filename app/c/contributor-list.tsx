import runtime from '@/adapter/effect/runtime'
import Co2AverageCmp from '@/components/co2-average'
import { type Co2Average, Co2Repository } from '@/domain/co2'
import { setLogLevelFromSearchParams } from '@/lib/utils'
import { Effect } from 'effect'
import type { SearchParams } from 'next/dist/server/request/search-params'
import Link from 'next/link'

import type { JSX } from 'react'

export default async function ContributorList(props: {
  searchParams: Promise<SearchParams>
}) {
  const searchParams = await props.searchParams
  return ContributorListEffect({ searchParams }).pipe(runtime.runPromise)
}

export function ContributorListEffect(props: { searchParams: SearchParams }) {
  return Co2Repository.getAllCo2Averages.pipe(
    Effect.map(filter(props.searchParams.search)),
    Effect.map(render),
    Effect.withSpan('Render /c/'),
    Effect.catchTag('DbError', handleDbError),
    setLogLevelFromSearchParams(props.searchParams),
  ) satisfies Effect.Effect<JSX.Element, never, unknown>
}

function render(co2Averages: Co2Average[]) {
  return (
    <>
      <ul className="grid grid-cols-[repeat(auto-fill,minmax(14rem,1fr))] gap-4">
        {co2Averages.map((co2Average) => (
          <Co2AverageCmp key={co2Average.slug} co2Average={co2Average} />
        ))}
      </ul>
      {co2Averages.length === 0 && (
        <div className="mx-auto max-w-xs text-foreground/50">
          <h2 className="mb-2 text-2xl">Nothing found üôÅ</h2>
          <p>
            Thank you for searching! Sadly this data is not available yet. Get
            in{' '}
            <Link className="underline" href="/about">
              contact
            </Link>
            , so that we can add more interesting data.
          </p>
        </div>
      )}
    </>
  )
}

function filter(term: string | string[] | null | undefined) {
  return (co2Averges: Co2Average[]) =>
    term && typeof term === 'string'
      ? co2Averges.filter((item) =>
          item.title.toLowerCase().includes(term.toLowerCase()),
        )
      : co2Averges
}

function handleDbError() {
  return Effect.succeed(<main>The database is not reachable...</main>)
}
