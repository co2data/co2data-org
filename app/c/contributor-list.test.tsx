import { Co2Average } from '@/domain/co2'
import { makeCo2Average } from '@/domain/co2/example-data'
import { Co2Repository } from '@/domain/co2/repository'
import { render, screen } from '@testing-library/react'
import { Effect, Layer, Option } from 'effect'
import { expect, test, vi } from 'vitest'
import { ContributorListEffect } from './contributor-list'

test('render no contributor found', async () => {
  const { container } = await setup({ mockData: [], searchParams: {} })
  expect(container).toMatchSnapshot()
})

test('render ', async () => {
  const searchParams = { search: 'pork' }

  const { container } = await setup({
    mockData: [makeCo2Average()],
    searchParams,
  })
  expect(container).toMatchSnapshot()
})

test('has link to pork', async () => {
  const searchParams = { search: 'pork' }

  await setup({ mockData: [makeCo2Average()], searchParams })
  const link = screen.getByRole('link', { name: 'Pork' })
  expect(link).toHaveAttribute('href', '/c/test')
})

async function setup(props: {
  mockData: Co2Average[]
  searchParams: { [key: string]: string | undefined }
}) {
  const co2RepoTest = Layer.succeed(
    Co2Repository,
    Co2Repository.of({
      getAllCo2Averages: () => Effect.succeed(props.mockData),
      getCo2AverageBySlug: (slug) => Effect.succeed(Option.none()),
    }),
  )
  const result = await ContributorListEffect(props).pipe(
    Effect.provide(co2RepoTest),
    Effect.runPromise,
  )

  return render(result)
}
