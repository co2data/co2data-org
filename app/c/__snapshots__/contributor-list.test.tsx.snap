// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`render  1`] = `
<div>
  <ul
    class="grid grid-cols-[repeat(auto-fill,_minmax(14rem,_1fr))] gap-8"
  >
    <li
      class="rounded border-4 border-border bg-card"
    >
      <p
        class="bg-border px-4 py-2 font-bold text-white"
      >
        <a
          href="/c/test"
        >
          Pork
        </a>
      </p>
      <p
        class="px-4 py-2 text-xs "
      >
        Possible CO
        <sub>
          2
        </sub>
        e per person-year
      </p>
      <p
        class="px-4 text-right"
      >
        <span
          class="text-4xl font-extrabold"
        >
          500
        </span>
        <span
          class="text-sm font-extrabold"
        >
          .
        </span>
        <span
          class="text-sm font-extrabold"
        >
          000
        </span>
         
        <span
          class="text-sm"
        >
          kg
        </span>
      </p>
      <div
        class="flex justify-between gap-2"
      >
        <p
          class="py-2 pl-4 text-xs"
        >
          1 
          kilogram
        </p>
        <p
          class="py-2 pr-4 text-xs"
        >
          30
           
          kg CO
          <sub>
            2
          </sub>
          e
        </p>
      </div>
    </li>
  </ul>
</div>
`;

exports[`render no contributor found 1`] = `
<div>
  <ul
    class="grid grid-cols-[repeat(auto-fill,_minmax(14rem,_1fr))] gap-8"
  />
  <div
    class="mx-auto max-w-xs text-foreground/50"
  >
    <h2
      class="mb-2 text-2xl"
    >
      Nothing found üôÅ
    </h2>
    <p>
      Thank you for searching! Sadly this data is not available yet. Get in
       
      <a
        class="underline"
        href="/about"
      >
        contact
      </a>
      , so that we can add more interesting data.
    </p>
  </div>
</div>
`;
